crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -1014,89 +1023,144 @@ struct RangeBlockMode<
             return Ok(Some(result));
         }
 
-        let Some(next_header) = self.iter.next() else {
-            return Ok(None);
-        };
+        // Next, try to get a completed task result if there are pending tasks
+        if !self.pending_tasks.is_empty() {
+            if let Some(task_result) = self.pending_tasks.next().await {
+                let mut chunk_results = task_result?;
+                if !chunk_results.is_empty() {
+                    let first_result = chunk_results.remove(0);
+                    self.next.extend(chunk_results);
+                    return Ok(Some(first_result));
+                }
+            }
+        }
 
-        let mut range_headers = Vec::with_capacity(self.max_range);
-        range_headers.push(next_header);
+        // If no pending tasks and no queued results, try to start new tasks
+        if self.pending_tasks.is_empty() {
+            // Try to get the next header(s) to process
+            let Some(next_header) = self.iter.next() else {
+                return Ok(None);
+            };
+
+            let mut range_headers = Vec::with_capacity(self.max_range);
+            range_headers.push(next_header);
+
+            // Collect consecutive blocks up to max_range size
+            while range_headers.len() < self.max_range {
+                let Some(peeked) = self.iter.peek() else { break };
+                let Some(last_header) = range_headers.last() else { break };
+
+                let expected_next = last_header.number() + 1;
+                if peeked.number() != expected_next {
+                    debug!(
+                        target: "rpc::eth::filter",
+                        last_block = last_header.number(),
+                        next_block = peeked.number(),
+                        expected = expected_next,
+                        "Non-consecutive block detected, stopping range collection"
+                    );
+                    break; // Non-consecutive block, stop here
+                }
 
-        // Collect consecutive blocks up to max_range size
-        while range_headers.len() < self.max_range {
-            let Some(peeked) = self.iter.peek() else { break };
-            let Some(last_header) = range_headers.last() else { break };
+                let Some(next_header) = self.iter.next() else { break };
+                range_headers.push(next_header);
+            }
 
-            let expected_next = last_header.header().number() + 1;
-            if peeked.header().number() != expected_next {
-                break; // Non-consecutive block, stop here
+            // Check if we should use parallel processing for large ranges
+            let remaining_headers = self.iter.len() + range_headers.len();
+            if remaining_headers >= PARALLEL_PROCESSING_THRESHOLD {
+                self.spawn_parallel_tasks(range_headers);
+            } else {
+                self.spawn_sequential_task(range_headers);
```

Comment:
here we'd need to await the sequential task directly and if it's parallel we we'd effectively loop or self.pending_tasks.next().await